nfs: add support for snapshot, restore and pvc-clone

Rakshith-R opened the pull request:
- 1d7191080 e2e: add tests for nfs snapshot and restore
- 3d6b57949 e2e: add pvc-pvc clone e2e test
- c44e83383 deploy: add nfs pvc-clone & pod-clone example yaml
- 827b14b8d nfs: add support for pvc-pvc clone
- f0b28b794 deploy: add support for nfs snapshot
- d7e4b910d nfs: add support for create & delete snapshot

Rakshith-R committed:
nfs: add support for create & delete snapshot

This commits adds support for creation and
deletion of nfs snapshots based on cephfs.

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R committed:
deploy: add support for nfs snapshot

This commit add external-snapshotter container and
required rbac to support nfs snaphots.
Example volumesnaphotclass, volumesnapshot,
pvc-restore and pod-restore yamls are also added.

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R committed:
nfs: add support for pvc-pvc clone

This commit adds support for pvc-pvc clone.
Only capability needed to be advertised, the
underlying support is already provided by cephfs
backend.

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R committed:
deploy: add nfs pvc-clone & pod-clone example yaml

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R committed:
e2e: add pvc-pvc clone e2e test

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R committed:
e2e: add tests for nfs snapshot and restore

Signed-off-by: Rakshith R <rar@redhat.com>

nixpanic reviewed (changes requested):
looks good to me, just need to update the README that includes the support for the new features

yati1998 reviewed (commented):
LGTM

Madhu-1 commented on deploy/nfs/kubernetes/csi-provisioner-rbac.yaml:
```diff
@@ -39,6 +39,21 @@ rules:
   - apiGroups: ["storage.k8s.io"]
     resources: ["csinodes"]
     verbs: ["get", "list", "watch"]
+  - apiGroups: ["snapshot.storage.k8s.io"]
+    resources: ["volumesnapshots"]
+    verbs: ["get", "list", "watch", "update", "patch"]
+  - apiGroups: ["snapshot.storage.k8s.io"]
+    resources: ["volumesnapshotcontents"]
+    verbs: ["create", "get", "list", "watch", "update", "delete", "patch"]
+  - apiGroups: ["snapshot.storage.k8s.io"]
+    resources: ["volumesnapshotclasses"]
+    verbs: ["get", "list", "watch"]
+  - apiGroups: ["snapshot.storage.k8s.io"]
+    resources: ["volumesnapshotcontents/status"]
+    verbs: ["update", "patch"]
+  - apiGroups: ["snapshot.storage.k8s.io"]
+    resources: ["volumesnapshots/status"]
+    verbs: ["update", "patch"]
```
Please reverify once do we need all these access?

Rakshith-R replied:
> Please reverify once do we need all these access?

okay, I am copying these permissions
https://github.com/kubernetes-csi/external-snapshotter/blob/v5.0.1/deploy/kubernetes/csi-snapshotter/rbac-csi-snapshotter.yaml#L33-L41

Madhu-1 commented on e2e/nfs.go:
```diff
@@ -569,6 +583,346 @@ var _ = Describe("nfs", func() {
+				opt := metav1.ListOptions{
+					LabelSelector: fmt.Sprintf("%s=%s", appKey, label[appKey]),
+				}
+				wErr := writeDataInPod(app, &opt, f)
```
use err here?

Madhu-1 replied:
writing data is required? if yes please add md5 checksum to make sure the content is matching.

Madhu-1 commented on e2e/nfs.go:
```diff
@@ -569,6 +583,346 @@ var _ = Describe("nfs", func() {
+				validateSubvolumeCount(f, parentPVCCount, fileSystemName, subvolumegroup)
+				validateOmapCount(f, parentPVCCount, cephfsType, metadataPool, volumesType)
+				validateOmapCount(f, totalCount, cephfsType, metadataPool, snapsType)
+				// create clones from different snapshots and bind it to an
+				// app
```
single-line comment should be fine?

Madhu-1 commented on e2e/nfs.go:
```diff
@@ -569,6 +583,346 @@ var _ = Describe("nfs", func() {
+				opt := metav1.ListOptions{
+					LabelSelector: fmt.Sprintf("%s=%s", appKey, label[appKey]),
+				}
+				wErr := writeDataInPod(app, &opt, f)
+				if wErr != nil {
+					e2elog.Failf("failed to write data from application %v", wErr)
+				}
+
```
writing data is required? if yes please add md5 checksum to make sure the content is matching.

Madhu-1 commented on e2e/snapshot.go:
```diff
@@ -206,6 +206,58 @@ func createCephFSSnapshotClass(f *framework.Framework) error {
+	if err != nil {
+		return err
+	}
+	_, err = sclient.VolumeSnapshotClasses().Create(context.TODO(), &sc, metav1.CreateOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to create volumesnapshotclass: %w", err)
+	}
```
extra create is not required as we are already creating in the PollImmediate loop?

Madhu-1 commented on e2e/snapshot.go:
```diff
@@ -206,6 +206,58 @@ func createCephFSSnapshotClass(f *framework.Framework) error {
+		return err
+	}
+
+	return sclient.VolumeSnapshotClasses().Delete(context.TODO(), sc.Name, metav1.DeleteOptions{})
```
use same retry mechanism to delete also?

Madhu-1 reviewed (commented):
no major comments, some nits. LGTM, as @nixpanic mentioned Doc update is required

Rakshith-R committed:
doc: mention support for nfs snapshot, restore & clone

NFS now supports snapshot, restore and pvc-pvc clone.

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R commented:
@mergifyio rebase

Madhu-1 commented on README.md:
```diff
@@ -114,6 +114,9 @@ for its support details.
 |        | Dynamically provision, de-provision File mode ROX volume  | Alpha          | >= v3.6.0          | >= v1.0.0        | Pacific (>=16.2.0)   | >= v1.14.0         |
 |        | Dynamically provision, de-provision File mode RWOP volume | Alpha          | >= v3.6.0          | >= v1.5.0        | Pacific (>=16.2.0)   | >= v1.22.0         |
 |        | Expand volume                                             | Alpha          | >= v3.7.0          | >= v1.1.0        | Pacific (>=16.2.0)   | >= v1.15.0         |
+|        | Creating and deleting snapshot                            | Alpha          | >= v3.7.0          | >= v1.1.0        | Pacific (>=16.2.0)   | >= v1.15.0         |
+|        | Provision volume from snapshot                            | Alpha          | >= v3.7.0          | >= v1.1.0        | Pacific (>=16.2.0)   | >= v1.15.0         |
+|        | Provision volume from another volume                      | Alpha          | >= v3.7.0          | >= v1.1.0        | Pacific (>=16.2.0)   | >= v1.15.0         |
```
reverify kubernetes support version it should be 1.17 for snapshot and 1.16 for PVC cone

Rakshith-R replied:
> reverify kubernetes support version it should be 1.17 for snapshot and 1.16 for PVC cone

missed it, using the correct version now, ptal

Madhu-1 reviewed (commented):
changes LGTM. small clarification is required.

nixpanic commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

nixpanic commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

nixpanic commented:
Both CI jobs failed with RBD image count out-of-sync

nixpanic commented:
**ci/centos/mini-e2e/k8s-1.22** failed with the new tests:
```
create a PVC clone and bind it to an app
...
May 18 11:52:43.119: INFO: waiting for PVC nfs-80811 (600 seconds elapsed)
May 18 11:52:43.121: INFO: failed to create PVC and app (nfs-80810): timed out waiting for the condition
May 18 11:52:43.121: INFO: failed to create PVC and app (nfs-80811): timed out waiting for the condition
May 18 11:52:43.121: INFO: failed to create PVC and app (nfs-80812): timed out waiting for the condition
May 18 11:52:43.121: FAIL: creating PVCs and apps failed, 3 errors were logged
```
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.22/detail/mini-e2e_k8s-1.22/3076/pipeline

Rakshith-R commented:
> **ci/centos/mini-e2e/k8s-1.22** failed with the new tests:
> 
> ```
> create a PVC clone and bind it to an app
> ...
> May 18 11:52:43.119: INFO: waiting for PVC nfs-80811 (600 seconds elapsed)
> May 18 11:52:43.121: INFO: failed to create PVC and app (nfs-80810): timed out waiting for the condition
> May 18 11:52:43.121: INFO: failed to create PVC and app (nfs-80811): timed out waiting for the condition
> May 18 11:52:43.121: INFO: failed to create PVC and app (nfs-80812): timed out waiting for the condition
> May 18 11:52:43.121: FAIL: creating PVCs and apps failed, 3 errors were logged
> ```
> 
> https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.22/detail/mini-e2e_k8s-1.22/3076/pipeline

```
May 18 11:52:43.138: INFO: At 2022-05-18 11:42:43 +0100 BST - event for nfs-80811: {persistentvolume-controller } ExternalProvisioning: waiting for a volume to be created, either by external provisioner "nfs.csi.ceph.com" or manually created by system administrator
May 18 11:52:43.138: INFO: At 2022-05-18 11:42:43 +0100 BST - event for nfs-80812: {nfs.csi.ceph.com_csi-nfsplugin-provisioner-b45fbb776-k7sm7_c5a32210-c1aa-42bd-a3de-6a4ef61f2e74 } ProvisioningFailed: failed to provision volume with StorageClass "csi-nfs-sc": error getting handle for DataSource Type VolumeSnapshot by Name nfs-80810: error getting snapshot nfs-80810 from api server: volumesnapshots.snapshot.storage.k8s.io "nfs-80810" is forbidden: User "system:serviceaccount:cephcsi-e2e-2a199dae:nfs-csi-provisioner" cannot get resource "volumesnapshots" in API group "snapshot.storage.k8s.io" in the namespace "nfs-8081"

```

external provisioner needs to access to volumesnapshot get/list to restore pvc from snapshot

Rakshith-R commented:
The mountPath did'nt match from pod.yaml to pod-clone.yaml and pod-restore.yaml
fixed it 
https://github.com/ceph/ceph-csi/compare/810832a49fbe149d281c8b8bff376b580ed421c0..fdfb0fe6a00f51fea2b73a45ecff43e87258f0b3
PTAL

Madhu-1 commented on deploy/nfs/kubernetes/csi-provisioner-rbac.yaml:
```diff
@@ -39,6 +39,18 @@ rules:
+    verbs: ["get", "list", "watch"]
+  - apiGroups: ["snapshot.storage.k8s.io"]
+    resources: ["volumesnapshotcontents"]
+    verbs: ["create", "get", "list", "watch", "update", "delete", "patch"]
```
create and delete access is not required remove it https://github.com/kubernetes-csi/external-snapshotter/blob/master/deploy/kubernetes/csi-snapshotter/rbac-csi-snapshotter.yaml#L37-L38?

Rakshith-R replied:
> create and delete access is not required remove it https://github.com/kubernetes-csi/external-snapshotter/blob/master/deploy/kubernetes/csi-snapshotter/rbac-csi-snapshotter.yaml#L37-L38?

release v5.0.1 which we are still using has this rbac, 
Let's update when we get v6.0.0 release

Rakshith-R commented:
https://github.com/ceph/ceph-csi/compare/2738cf9ac4d78591316d159b577dfa24e8b47ebf..a5a3e2a42afe1b39fc0a7f0996be07e68d7bd389

only parent PVC had to be deleted. 
demo pod had been deleted during call to 
https://github.com/ceph/ceph-csi/blob/7e1ed75cfbc0ba4465b1945c4663f95934f773ac/e2e/utils.go#L772

Rakshith-R commented:
/retest all

Rakshith-R committed:
ci: increase e2e_timeout to 150 mins

With the added nfs snap and clone tests,
e2e test is timing out. This commit increases
the timeout to 150 mins.

Signed-off-by: Rakshith R <rar@redhat.com>

Rakshith-R commented:
:partying_face: ci is all green, PTAL

https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.23/detail/mini-e2e_k8s-1.23/1748/pipeline/89
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.22/detail/mini-e2e_k8s-1.22/3114/pipeline/89

Rakshith-R commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22
ci passed but
jenkins is not updating the status :disappointed: 
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.22/detail/mini-e2e-helm_k8s-1.22/3225/pipeline

Rakshith-R commented:
/retest ci/centos/mini-e2e-helm/k8s-1.22

Rakshith-R commented:
@nixpanic @yati1998 can you please take a look?

Rakshith-R commented:
@mergifyio rebase
